Support new Firmware of Pylontech Battery

<!-- READ THIS FIRST:
 - This is for feature requests only, for issues please go to the issues repository.
 - Please be as descriptive as possible, especially use-cases that can otherwise not be solved boost the problem's priority.

 DO NOT DELETE ANY TEXT from this template! Otherwise the issue may be closed without a comment.
-->

**Describe the problem you have/What new integration you would like**
<!-- A description of what you want to happen. -->
Pylontech has change in a Firmwareupdate the Table after the command "pwr" 

I think we need an update of this integration to use both Firmware. 

**Please describe your use case for this integration and alternatives you've tried:**
<!-- A clear and concise description of what the problem is. -->
The Pylontech intergration does not work with ny Battery. 
I use this intergration to get the data over MQTT: 
https://github.com/hidaba/PylontechMonitoring
I change the parce of the Value (pLineStart) to the new values now it work 


**Additional context**
<!-- Add any other context about the feature request here. -->

The Output of the command "pwr" of my Battery: 

`pwr

@

Power Volt   Curr   Tempr  Tlow   Tlow.Id  Thigh  Thigh.Id Vlow   Vlow.Id  Vhigh  Vhigh.Id Base.St  Volt.St  Curr.St  Temp.St  Coulomb  Time                 B.V.St   B.T.St  MosTempr M.T.St   SysAlarm.St

1     48844  -33847 17300  16300  12       17200  0        3251   13       3260   8        Dischg   Normal   Normal   Normal   89%      2025-04-07 02:01:11  Normal   Normal  17700    Normal   Normal  

2     48895  -22701 18900  17200  10       17700  5        3255   0        3262   2        Dischg   Normal   Normal   Normal   84%      2025-04-07 02:01:08  Normal   Normal  18100    Normal   Normal  

3     48889  -22950 19200  17600  10       17900  5        3254   5        3262   2        Dischg   Normal   Normal   Normal   82%      2025-04-07 02:01:08  Normal   Normal  18900    Normal   Normal  

4     -      -      -      -      -      -      -      Absent   -        -        -        -        -                    -        -       

5     -      -      -      -      -      -      -      Absent   -        -        -        -        -                    -        -       

6     -      -      -      -      -      -      -      Absent   -        -        -        -        -                    -        -       

7     -      -      -      -      -      -      -      Absent   -        -        -        -        -                    -        -       

8     -      -      -      -      -      -      -      Absent   -        -        -        -        -                    -        -       

9     -      -      -      -      -      -      -      Absent   -        -        -        -        -                    -        -       

10    -      -      -      -      -      -      -      Absent   -        -        -        -        -                    -        -       

11    -      -      -      -      -      -      -      Absent   -        -        -        -        -                    -        -       

12    -      -      -      -      -      -      -      Absent   -        -        -        -        -                    -        -       

13    -      -      -      -      -      -      -      Absent   -        -        -        -        -                    -        -       

14    -      -      -      -      -      -      -      Absent   -        -        -        -        -                    -        -       

15    -      -      -      -      -      -      -      Absent   -        -        -        -        -                    -        -       

16    -      -      -      -      -      -      -      Absent   -        -        -        -        -                    -        -       

Command completed successfully

$$

pylon_debug>`

New is 

Tlow.Id  
Thigh.Id 
Vlow.Id  
Vhigh.Id 
SysAlarm.St

My Battery Info

`info

@

Device address      : 1

Manufacturer        : Pylon

Device name         : US5000

Board version       : V10R04

Board               : NF4.E3

Main Soft version   : B69.25.0.0

Soft  version       : V2.0

Boot  version       : V1.0

Comm version        : V2.0

Release Date        : 24-05-13

Barcode             :  --deleted--



Specification       : 48V/100AH

Cell Number         : 15

Max Dischg Curr     : -105000mA

Max Charge Curr     : 105000mA

EPONPort rate       : 1200

Console Port rate   : 115200

Customer Ver.ID     : STD   

Command completed successfully

$$

pylon_debug>`


<!-- LEAVE THIS LINE AS-IS AND DON'T DELETE IT, OTHERWISE THE FEATURE REQUEST WILL BE CLOSED AUTOMATICALLY. -->

Hi and thank you very much. I am happy to test the changes.

I get an error while compiling.

```
src/esphome/components/pylontech/pylontech.cpp: In lambda function:
src/esphome/components/pylontech/pylontech.cpp:112:30: error: 'min' was not declared in this scope
  112 |     memcpy(token_buf, start, min(cursor - start, TEXT_SENSOR_MAX_LEN - 1));
      |                              ^~~
src/esphome/components/pylontech/pylontech.cpp:112:30: note: suggested alternatives:
In file included from /home/ar/.platformio/packages/toolchain-riscv32-esp/riscv32-esp-elf/include/c++/14.2.0/algorithm:61,
                 from src/esphome/core/optional.h:19,
                 from src/esphome/core/component.h:9,
                 from src/esphome/components/pylontech/pylontech.h:3,
                 from src/esphome/components/pylontech/pylontech.cpp:1:
/home/ar/.platformio/packages/toolchain-riscv32-esp/riscv32-esp-elf/include/c++/14.2.0/bits/stl_algo.h:5695:5: note:   'std::min'
 5695 |     min(initializer_list<_Tp> __l, _Compare __comp)
      |     ^~~
In file included from /home/ar/.platformio/packages/toolchain-riscv32-esp/riscv32-esp-elf/include/c++/14.2.0/tuple:44,
                 from /home/ar/.platformio/packages/toolchain-riscv32-esp/riscv32-esp-elf/include/c++/14.2.0/functional:53,
                 from src/esphome/core/component.h:5:
/home/ar/.platformio/packages/toolchain-riscv32-esp/riscv32-esp-elf/include/c++/14.2.0/bits/ranges_util.h:754:29: note:   'std::ranges::min'
  754 |   inline constexpr __min_fn min{};
      |                             ^~~
Compiling .pioenvs/eth01-evo-poe/src/esphome/components/safe_mode/safe_mode.cpp.o
*** [.pioenvs/eth01-evo-poe/src/esphome/components/pylontech/pylontech.cpp.o] Error 1
```

Some notes:
I use a ESP32 C3 wird `esp-idf` framework (I can't use Arduino)

--- a/esphome/components/pylontech/pylontech.cpp
+++ b/esphome/components/pylontech/pylontech.cpp
@@ -16,7 +16,14 @@
   }
 
 #define PARSE_STR(field, field_name) \
-  { get_token(field); }
+  { \
+    get_token(field); \
+    if (strlen(field) < 2) { \
+      ESP_LOGD(TAG, "too short " field_name " in line %s", buffer.substr(0, buffer.size() - 2).c_str()); \
+      return; \
+    } \
+  }
+
 namespace esphome {
 namespace pylontech {
 
@@ -112,8 +119,9 @@ void PylontechComponent::process_line_(std::string &buffer) {
       cursor++;
     }
 
-    memcpy(token_buf, start, std::min(cursor - start, TEXT_SENSOR_MAX_LEN - 1));
-    token_buf[std::min((cursor - start), TEXT_SENSOR_MAX_LEN - 1)] = 0;
+    size_t token_len = std::min(static_cast<size_t>(cursor - start), static_cast<size_t>(TEXT_SENSOR_MAX_LEN - 1));
+    memcpy(token_buf, start, token_len);
+    token_buf[token_len] = 0;
   };
 
   {
@@ -188,3 +196,6 @@ float PylontechComponent::get_setup_priority() const { return setup_priority::DA
 
 }  // namespace pylontech
 }  // namespace esphome
+
+#undef PARSE_INT
+#undef PARSE_STR

--- a/esphome/components/pylontech/pylontech.cpp
+++ b/esphome/components/pylontech/pylontech.cpp
@@ -83,6 +83,9 @@ void PylontechComponent::process_line_(std::string &buffer) {
   // Power Volt   Curr   Tempr  Tlow   Thigh  Vlow   Vhigh  Base.St  Volt.St  Curr.St  Temp.St  Coulomb  Time                 B.V.St   B.T.St   MosTempr M.T.St
   // 1     50548  8910   25000  24200  25000  3368   3371   Charge   Normal   Normal   Normal   97%      2021-06-30 20:49:45  Normal  Normal  22700    Normal
   // 1     46012  1255   9100   5300   5500   3047   3091   SysError Low      Normal   Normal   4%       2025-11-28 17:56:33  Low      Normal  7800     Normal
+  // newer firmware example:
+  // Power Volt Curr Tempr Tlow Tlow.Id Thigh Thigh.Id Vlow Vlow.Id Vhigh Vhigh.Id Base.St Volt.St Curr.St Temp.St Coulomb Time                B.V.St B.T.St MosTempr M.T.St SysAlarm.St
+  // 1     49405 0   17600 13700 8      14500 0        3293 2       3294   0       Idle    Normal  Normal  Normal  60%     2025-12-05 00:53:41 Normal Normal 16600    Normal Normal
   // clang-format on
 
   PylontechListener::LineContents l{};

--- a/esphome/components/pylontech/pylontech.cpp
+++ b/esphome/components/pylontech/pylontech.cpp
@@ -2,6 +2,21 @@
 #include "esphome/core/helpers.h"
 #include "esphome/core/log.h"
 
+// Helper macros
+#define PARSE_INT(field, field_name) \
+  { \
+    get_token(token_buf); \
+    auto val = parse_number<int>(token_buf); \
+    if (val.has_value()) { \
+      field = val.value(); \
+    } else { \
+      ESP_LOGD(TAG, "invalid " field_name " in line %s", buffer.substr(0, buffer.size() - 2).c_str()); \
+      return; \
+    } \
+  }
+
+#define PARSE_STR(field, field_name) \
+  { get_token(field); }
 namespace esphome {
 namespace pylontech {
 
@@ -64,33 +79,102 @@ void PylontechComponent::loop() {
 void PylontechComponent::process_line_(std::string &buffer) {
   ESP_LOGV(TAG, "Read from serial: %s", buffer.substr(0, buffer.size() - 2).c_str());
   // clang-format off
-  // example line to parse:
-  // Power Volt  Curr Tempr Tlow  Thigh  Vlow Vhigh Base.St Volt.St Curr.St Temp.St Coulomb Time                B.V.St B.T.St MosTempr M.T.St
-  // 1    50548  8910 25000 24200 25000  3368 3371  Charge  Normal  Normal  Normal  97%     2021-06-30 20:49:45 Normal Normal 22700    Normal
+  // example lines to parse:
+  // Power Volt   Curr   Tempr  Tlow   Thigh  Vlow   Vhigh  Base.St  Volt.St  Curr.St  Temp.St  Coulomb  Time                 B.V.St   B.T.St   MosTempr M.T.St
+  // 1     50548  8910   25000  24200  25000  3368   3371   Charge   Normal   Normal   Normal   97%      2021-06-30 20:49:45  Normal  Normal  22700    Normal
+  // 1     46012  1255   9100   5300   5500   3047   3091   SysError Low      Normal   Normal   4%       2025-11-28 17:56:33  Low      Normal  7800     Normal
   // clang-format on
 
   PylontechListener::LineContents l{};
-  char mostempr_s[6];
-  const int parsed = sscanf(                                                                                   // NOLINT
-      buffer.c_str(), "%d %d %d %d %d %d %d %d %7s %7s %7s %7s %d%% %*d-%*d-%*d %*d:%*d:%*d %*s %*s %5s %*s",  // NOLINT
-      &l.bat_num, &l.volt, &l.curr, &l.tempr, &l.tlow, &l.thigh, &l.vlow, &l.vhigh, l.base_st, l.volt_st,      // NOLINT
-      l.curr_st, l.temp_st, &l.coulomb, mostempr_s);                                                           // NOLINT
 
-  if (l.bat_num <= 0) {
-    ESP_LOGD(TAG, "invalid bat_num in line %s", buffer.substr(0, buffer.size() - 2).c_str());
-    return;
+  const char *cursor = buffer.c_str();
+  char token_buf[TEXT_SENSOR_MAX_LEN] = {0};
+
+  // Helper Lambda to extract tokens
+  auto get_token = [&](char *token_buf) -> void {
+    // Skip leading whitespace
+    while (*cursor == ' ' || *cursor == '\t') {
+      cursor++;
+    }
+
+    if (*cursor == '\0') {
+      token_buf[0] = 0;
+      return;
+    }
+
+    const char *start = cursor;
+
+    // Find end of field
+    while (*cursor != '\0' && *cursor != ' ' && *cursor != '\t' && *cursor != '\r') {
+      cursor++;
+    }
+
+    memcpy(token_buf, start, std::min(cursor - start, TEXT_SENSOR_MAX_LEN - 1));
+    token_buf[std::min((cursor - start), TEXT_SENSOR_MAX_LEN - 1)] = 0;
+  };
+
+  {
+    get_token(token_buf);
+    auto val = parse_number<int>(token_buf);
+    if (val.has_value() && val.value() > 0) {
+      l.bat_num = val.value();
+    } else if (strcmp(token_buf, "Power") == 0) {
+      // header line i.e. "Power Volt   Curr" and so on
+      this->has_tlow_id = buffer.find("Tlow.Id") != std::string::npos;
+      ESP_LOGD(TAG, "header line %s Tlow.Id: %s", this->has_tlow_id ? "with" : "without",
+               buffer.substr(0, buffer.size() - 2).c_str());
+      return;
+    } else {
+      ESP_LOGD(TAG, "unknown line %s", buffer.substr(0, buffer.size() - 2).c_str());
+      return;
+    }
   }
-  if (parsed != 14) {
-    ESP_LOGW(TAG, "invalid line: found only %d items in %s", parsed, buffer.substr(0, buffer.size() - 2).c_str());
-    return;
+  PARSE_INT(l.volt, "Volt");
+  PARSE_INT(l.curr, "Curr");
+  PARSE_INT(l.tempr, "Tempr");
+  PARSE_INT(l.tlow, "Tlow");
+  if (this->has_tlow_id) {
+    get_token(token_buf);  // Skip Tlow.Id
   }
-  auto mostempr_parsed = parse_number<int>(mostempr_s);
-  if (mostempr_parsed.has_value()) {
-    l.mostempr = mostempr_parsed.value();
-  } else {
-    l.mostempr = -300;
-    ESP_LOGW(TAG, "bat_num %d: received no mostempr", l.bat_num);
+  PARSE_INT(l.thigh, "Thigh");
+  if (this->has_tlow_id) {
+    get_token(token_buf);  // Skip Thigh.Id
   }
+  PARSE_INT(l.vlow, "Vlow");
+  if (this->has_tlow_id) {
+    get_token(token_buf);  // Skip Vlow.Id
+  }
+  PARSE_INT(l.vhigh, "Vhigh");
+  if (this->has_tlow_id) {
+    get_token(token_buf);  // Skip Vhigh.Id
+  }
+  PARSE_STR(l.base_st, "Base.St");
+  PARSE_STR(l.volt_st, "Volt.St");
+  PARSE_STR(l.curr_st, "Curr.St");
+  PARSE_STR(l.temp_st, "Temp.St");
+  {
+    get_token(token_buf);
+    for (int i = 0; i < TEXT_SENSOR_MAX_LEN; i++) {
+      if (token_buf[i] == '%') {
+        token_buf[i] = 0;
+        break;
+      }
+    }
+    auto coul_val = parse_number<int>(token_buf);
+    if (coul_val.has_value()) {
+      l.coulomb = coul_val.value();
+    } else {
+      ESP_LOGD(TAG, "invalid Coulomb in line %s", buffer.substr(0, buffer.size() - 2).c_str());
+      return;
+    }
+  }
+  get_token(token_buf);  // Skip Date
+  get_token(token_buf);  // Skip Time
+  get_token(token_buf);  // Skip B.V.St
+  get_token(token_buf);  // Skip B.T.St
+  PARSE_INT(l.mostempr, "Mostempr");
+
+  ESP_LOGD(TAG, "successful line %s", buffer.substr(0, buffer.size() - 2).c_str());
 
   for (PylontechListener *listener : this->listeners_) {
     listener->on_line_read(&l);

--- a/esphome/components/pylontech/pylontech.cpp
+++ b/esphome/components/pylontech/pylontech.cpp
@@ -8,7 +8,7 @@
     get_token(token_buf); \
     auto val = parse_number<int>(token_buf); \
     if (val.has_value()) { \
-      field = val.value(); \
+      (field) = val.value(); \
     } else { \
       ESP_LOGD(TAG, "invalid " field_name " in line %s", buffer.substr(0, buffer.size() - 2).c_str()); \
       return; \
@@ -123,8 +123,8 @@ void PylontechComponent::process_line_(std::string &buffer) {
       l.bat_num = val.value();
     } else if (strcmp(token_buf, "Power") == 0) {
       // header line i.e. "Power Volt   Curr" and so on
-      this->has_tlow_id = buffer.find("Tlow.Id") != std::string::npos;
-      ESP_LOGD(TAG, "header line %s Tlow.Id: %s", this->has_tlow_id ? "with" : "without",
+      this->has_tlow_id_ = buffer.find("Tlow.Id") != std::string::npos;
+      ESP_LOGD(TAG, "header line %s Tlow.Id: %s", this->has_tlow_id_ ? "with" : "without",
                buffer.substr(0, buffer.size() - 2).c_str());
       return;
     } else {
@@ -136,19 +136,19 @@ void PylontechComponent::process_line_(std::string &buffer) {
   PARSE_INT(l.curr, "Curr");
   PARSE_INT(l.tempr, "Tempr");
   PARSE_INT(l.tlow, "Tlow");
-  if (this->has_tlow_id) {
+  if (this->has_tlow_id_) {
     get_token(token_buf);  // Skip Tlow.Id
   }
   PARSE_INT(l.thigh, "Thigh");
-  if (this->has_tlow_id) {
+  if (this->has_tlow_id_) {
     get_token(token_buf);  // Skip Thigh.Id
   }
   PARSE_INT(l.vlow, "Vlow");
-  if (this->has_tlow_id) {
+  if (this->has_tlow_id_) {
     get_token(token_buf);  // Skip Vlow.Id
   }
   PARSE_INT(l.vhigh, "Vhigh");
-  if (this->has_tlow_id) {
+  if (this->has_tlow_id_) {
     get_token(token_buf);  // Skip Vhigh.Id
   }
   PARSE_STR(l.base_st, "Base.St");
@@ -157,9 +157,9 @@ void PylontechComponent::process_line_(std::string &buffer) {
   PARSE_STR(l.temp_st, "Temp.St");
   {
     get_token(token_buf);
-    for (int i = 0; i < TEXT_SENSOR_MAX_LEN; i++) {
-      if (token_buf[i] == '%') {
-        token_buf[i] = 0;
+    for (char &i : token_buf) {
+      if (i == '%') {
+        i = 0;
         break;
       }
     }

--- a/esphome/components/pylontech/pylontech.h
+++ b/esphome/components/pylontech/pylontech.h
@@ -8,14 +8,14 @@ namespace esphome {
 namespace pylontech {
 
 static const uint8_t NUM_BUFFERS = 20;
-static const uint8_t TEXT_SENSOR_MAX_LEN = 8;
+static const uint8_t TEXT_SENSOR_MAX_LEN = 14;
 
 class PylontechListener {
  public:
   struct LineContents {
     int bat_num = 0, volt, curr, tempr, tlow, thigh, vlow, vhigh, coulomb, mostempr;
-    char base_st[TEXT_SENSOR_MAX_LEN], volt_st[TEXT_SENSOR_MAX_LEN], curr_st[TEXT_SENSOR_MAX_LEN],
-        temp_st[TEXT_SENSOR_MAX_LEN];
+    char base_st[TEXT_SENSOR_MAX_LEN] = {0}, volt_st[TEXT_SENSOR_MAX_LEN] = {0}, curr_st[TEXT_SENSOR_MAX_LEN] = {0},
+         temp_st[TEXT_SENSOR_MAX_LEN] = {0};
   };
 
   virtual void on_line_read(LineContents *line);
@@ -45,6 +45,7 @@ class PylontechComponent : public PollingComponent, public uart::UARTDevice {
   std::string buffer_[NUM_BUFFERS];
   int buffer_index_write_ = 0;
   int buffer_index_read_ = 0;
+  bool has_tlow_id = false;
 
   std::vector<PylontechListener *> listeners_{};
 };

--- a/esphome/components/pylontech/pylontech.h
+++ b/esphome/components/pylontech/pylontech.h
@@ -45,7 +45,7 @@ class PylontechComponent : public PollingComponent, public uart::UARTDevice {
   std::string buffer_[NUM_BUFFERS];
   int buffer_index_write_ = 0;
   int buffer_index_read_ = 0;
-  bool has_tlow_id = false;
+  bool has_tlow_id_ = false;
 
   std::vector<PylontechListener *> listeners_{};
 };

